File: LR1_GO/Queue.go
package main

import "fmt"

func QPUSH(queue *Queue, value string) {
	newNode := &QNode{Data: value, Next: nil}
	if queue.Rear == nil {
		queue.Front = newNode
		queue.Rear = newNode
	} else {
		queue.Rear.Next = newNode
		queue.Rear = newNode
	}
	queue.Size++
}

func QPOP(queue *Queue) {
	if queue.Front == nil {
		fmt.Println("Очередь пуста!")
		return
	}
	
	queue.Front = queue.Front.Next
	if queue.Front == nil {
		queue.Rear = nil
	}
	queue.Size--
}

func QPRINT(queue *Queue) {
	if queue.Front == nil {
		fmt.Println("Очередь пуста!")
		return
	}
	
	fmt.Printf("Очередь (%d): ", queue.Size)
	curr := queue.Front
	for curr != nil {
		fmt.Print(curr.Data)
		if curr.Next != nil {
			fmt.Print(" -> ")
		}
		curr = curr.Next
	}
	fmt.Println()
}

func QCLEAR(queue *Queue) {
	queue.Front = nil
	queue.Rear = nil
	queue.Size = 0
	fmt.Println("Очередь очищена")
}

File: LR1/Queue.cpp
#include "Queue.h"
#include <iostream>
#include <fstream>

using namespace std;

void QPUSH(Queue& queue, const string& value) {
    QNode* newNode = new QNode(value, nullptr);
    if (!queue.rear) {
        queue.front = queue.rear = newNode;
    } else {
        queue.rear->next = newNode;
        queue.rear = newNode;
    }
    queue.size++;    
}

void QPOP(Queue& queue) {
    if (!queue.front) { 
        cout << "Очередь пуста!\n"; 
        return; 
    }
    QNode* temp = queue.front;
    queue.front = queue.front->next;
    if (!queue.front) queue.rear = nullptr;
    delete temp;
    queue.size--;
}

void QPRINT(const Queue& queue) {
    if (!queue.front) { 
        cout << "Очередь пуста!\n"; 
        return; 
    }
    cout << "Очередь (" << queue.size << "): ";
    QNode* curr = queue.front;
    while (curr) {
        cout << curr->data;
        if (curr->next) cout << " -> ";
        curr = curr->next;
    }
    cout << endl;
}

void QCLEAR(Queue& queue) {
    while (queue.front) {
        QNode* temp = queue.front;
        queue.front = queue.front->next;
        delete temp;
    }
    queue.rear = nullptr;
    queue.size = 0;
    cout << "Очередь очищена" << endl;
}
